### ISSUE TITLE
Adapt to "More explicit foralls" on GHC HEAD

### ISSUE THREAD

**Issue by RyanGlScott** [glguy/th-abstraction#55]
GHC HEAD just landed the commit [`More explicit foralls (GHC Proposal 0007)`](http://git.haskell.org/ghc.git/commit/512eeb9bb9a81e915bfab25ca16bc87c62252064), which extends the Template Haskell AST forms for data family instances with explicit `forall`s. In other words, it applies this change:

```diff
diff --git a/libraries/template-haskell/Language/Haskell/TH/Syntax.hs b/libraries/template-haskell/Language/Haskell/TH/Syntax.hs
index 7ee81c8..b75a048 100644 (file)
--- a/libraries/template-haskell/Language/Haskell/TH/Syntax.hs
+++ b/libraries/template-haskell/Language/Haskell/TH/Syntax.hs
@@ -1711,14 +1711,18 @@ data Dec
                (Maybe Kind)
          -- ^ @{ data family T a b c :: * }@
 
-  | DataInstD Cxt Name [Type]
+  | DataInstD Cxt Name
+             (Maybe [TyVarBndr])  -- Quantified type vars
+             [Type]
              (Maybe Kind)         -- Kind signature
              [Con] [DerivClause]  -- ^ @{ data instance Cxt x => T [x]
                                   --       = A x | B (T x)
                                   --       deriving (Z,W)
                                   --       deriving stock Eq }@
 
-  | NewtypeInstD Cxt Name [Type]
+  | NewtypeInstD Cxt Name
+                 (Maybe [TyVarBndr])  -- Quantified type vars
+                 [Type]
                  (Maybe Kind)      -- Kind signature
                  Con [DerivClause] -- ^ @{ newtype instance Cxt x => T [x]
                                    --        = A (B x)
```

One important practical consequence of this change is that `th-abstraction` now fails to build on GHC HEAD:

```
[2 of 2] Compiling Language.Haskell.TH.Datatype ( src/Language/Haskell/TH/Datatype.hs, /home/rgscott/Documents/Hacking/Haskell/th-abstraction/dist-newstyle/build/x86_64-linux/ghc-8.7.20181027/th-abstraction-0.2.8.0/build/Language/Haskell/TH/Datatype.o )

src/Language/Haskell/TH/Datatype.hs:525:9: error:
    • The constructor ‘NewtypeInstD’ should have 7 arguments, but has been given 6
    • In the pattern: NewtypeInstD cx n ts k c deriv
      In a stmt of a pattern guard for
                     an equation for ‘repairDataFam’:
        NewtypeInstD cx n ts k c deriv <- instD
      In an equation for ‘repairDataFam’:
          repairDataFam famD instD
            | DataFamilyD _ dvars _ <- famD,
              NewtypeInstD cx n ts k c deriv <- instD
            = NewtypeInstD cx n (repairVarKindsWith dvars ts) k c deriv
            | DataFamilyD _ dvars _ <- famD,
              DataInstD cx n ts k c deriv <- instD
            = DataInstD cx n (repairVarKindsWith dvars ts) k c deriv
    |
525 |       , NewtypeInstD cx n ts k c deriv <- instD
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Language/Haskell/TH/Datatype.hs:579:5: error:
    • The constructor ‘NewtypeInstD’ should have 7 arguments, but has been given 6
    • In the pattern:
        NewtypeInstD context name params _kind con _derives
      In a case alternative:
          NewtypeInstD context name params _kind con _derives
            -> repair13618' . giveTypesStarKinds
                 =<<
                   normalizeDec' isReified context name params [con] NewtypeInstance
      In the expression:
        case dec of
          NewtypeD context name tyvars _kind con _derives
            -> giveTypesStarKinds
                 <$>
                   normalizeDec'
                     isReified context name (bndrParams tyvars) [con] Newtype
          DataD context name tyvars _kind cons _derives
            -> giveTypesStarKinds
                 <$>
                   normalizeDec'
                     isReified context name (bndrParams tyvars) cons Datatype
          NewtypeInstD context name params _kind con _derives
            -> repair13618' . giveTypesStarKinds
                 =<<
                   normalizeDec' isReified context name params [con] NewtypeInstance
          DataInstD context name params _kind cons _derives
            -> repair13618' . giveTypesStarKinds
                 =<< normalizeDec' isReified context name params cons DataInstance
          _ -> fail "normalizeDecFor: DataD or NewtypeD required"
    |
579 |     NewtypeInstD context name params _kind con _derives ->
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Language/Haskell/TH/Datatype.hs:1638:45: error:
    • Couldn't match type ‘Type -> TySynEqn’ with ‘TySynEqn’
      Expected type: Q TySynEqn
        Actual type: Q (Type -> TySynEqn)
    • In the second argument of ‘(<$>)’, namely
        ‘(TySynEqn <$> sequence ps <*> r)’
      In the expression:
        TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
      In an equation for ‘tySynInstDCompat’:
          tySynInstDCompat n ps r
            = TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
     |
1638 | tySynInstDCompat n ps r = TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
     |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Language/Haskell/TH/Datatype.hs:1638:58: error:
    • Couldn't match type ‘[]’ with ‘Maybe’
      Expected type: Q (Maybe [TyVarBndr])
        Actual type: Q [Type]
    • In the second argument of ‘(<$>)’, namely ‘sequence ps’
      In the first argument of ‘(<*>)’, namely ‘TySynEqn <$> sequence ps’
      In the second argument of ‘(<$>)’, namely
        ‘(TySynEqn <$> sequence ps <*> r)’
     |
1638 | tySynInstDCompat n ps r = TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
     |                                                          ^^^^^^^^^^^

src/Language/Haskell/TH/Datatype.hs:1638:74: error:
    • Couldn't match type ‘Type’ with ‘[Type]’
      Expected type: Q [Type]
        Actual type: TypeQ
    • In the second argument of ‘(<*>)’, namely ‘r’
      In the second argument of ‘(<$>)’, namely
        ‘(TySynEqn <$> sequence ps <*> r)’
      In the expression:
        TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
     |
1638 | tySynInstDCompat n ps r = TySynInstD n <$> (TySynEqn <$> sequence ps <*> r)
     |                                                                          ^
```

Of course, we could just work around this with some more CPP. But now that GHC presents us with this information, I wonder if `th-abstraction` should take advantage of it.

In particular, I've always been slightly unsatisfied with the way `th-abstraction` represents data family instances. To explain what I mean, consider the definition of `DatatypeInfo`:

```haskell
data DatatypeInfo = DatatypeInfo
  { datatypeContext :: Cxt               -- ^ Data type context (deprecated)
  , datatypeName    :: Name              -- ^ Type constructor
  , datatypeVars    :: [Type]            -- ^ Type parameters
  , datatypeVariant :: DatatypeVariant   -- ^ Extra information
  , datatypeCons    :: [ConstructorInfo] -- ^ Normalize constructor information
  }
```

What are the type variables bound by a `DatatypeInfo`? For a normal data type, they're simply the `dataypeVars` (although they're represented as `Type`s instead of `TyVarBndr`s). But for a data family instance, it's not as straightforward to determine this, since `datatypeVars` can be arbitrary types, such as in this example:

```haskell
data family Foo x y

data instance Foo (Maybe a) (Either a b) = MkFoo a b
```

Here, we have `datatypeVars = [Maybe a, Either a b]`. One can try to figure out the bound type variables from `datatypeVars` by doing some _post facto_ analysis, but this isn't always possible (see [Trac #13809](https://ghc.haskell.org/trac/ghc/ticket/13809), for example).

What I think would be a nicer story is if we explicitly stated what the bound type variables were directly in `DatatypeInfo`. That is to say, add an additional field to `DatatypeInfo` that looks something like:

```haskell
datatypeTyVarBndrs :: [TyVarBndr] -- ^ Bound type variables
```

Thanks to the [`More explicit foralls (GHC Proposal 0007)`](http://git.haskell.org/ghc.git/commit/512eeb9bb9a81e915bfab25ca16bc87c62252064) commit, we now have enough information to always fill this field in accurately, regardless of whether we're dealing with a normal data type or a data family instance. On older GHCs, we'll have to fall back to using the _post facto_ analysis I mentioned above, but them's the breaks.

One awkward part of this plan is that if we add a `datatypeTyVarBndrs` field, the name of the existing `datatypeVars :: [Type]` field will become very misleading. Perhaps we ought to rename `datatypeVars` to something else? (In fact, `datatypeVars` might be a more suitable name for the new field of type `[TyVarBndr]`.)

What do you think about this plan, @glguy?

**Comment by glguy** [comment]
I think it makes sense to improve things with data families in this direction. Renaming the existing field makes sense to me, too. I think that this package has few enough users that we can afford to simply make the best design choices and then update downstream code accordingly.

**Comment by RyanGlScott** [comment]
Fixed in #65.
